ContextStack as hybrid stack

lontivero opened the pull request:
This PR is for improving the ```ScriptEvaluationContext```'s stack operations, because the ```System.Collections.Generic.Stack``` class didn't fit very well. Basically, it is necessary to perform some operations that are not provided by the Stack class:

* Insert, 
* Swap,
* Erase (renamed as Remove) and,
* Top

This PR includes a new ```ContextStack``` stack class that provides these missing methods and operates in the internal array. 
 
For example:

```c#
    swap(ref _Stack, -3, -2);   ---> _stack.Swap(-3, -2);
``` 

As a side product, there is a very small gain in performance.

Please @NicolasDorier, take a look at these changes and give me some feedback.

Ah! there was a broken test in the RestClientTests because someone spent an UTXO that I used in the test (now, the test uses an UTXO that is mine and that I won't spend)

NicolasDorier commented:
Not sure DebuggerTypeProxy will not crash on other plateform.

You also wrote "tipically" instead of "typically".

The PR messed up indentation which make it hard to review, but it seems right.

lontivero commented:
> Not sure DebuggerTypeProxy will not crash on other plateform.

I have no the answer but I think it shouldn't. I removed it from the PR.

> You also wrote "tipically" instead of "typically".

Fixed

> The PR messed up indentation which make it hard to review, but it seems right.

I know. I made so many changes and then rolled them back that the code changed a lot and messed it up. I'm sorry for that. But the changes are really a few.
